### PR TITLE
DNM: pybind/mgr: update smart mgr module

### PR DESCRIPTION
**Pull request by yaarith**
smart module now can periodically scrape SMART data from disk, run
failure predictions, and warn/mark out OSDs which use 'about to fail'
devices. Currently supports only SATA disks.

Signed-off-by: Yaarit Hatuka yaarithatuka@gmail.com

### CODE REVIEW

File: src/pybind/mgr/smart/module.py [hunk 1]
```diff
@@ -1,33 +1,561 @@
...
+                if data['hostname'] == host:
+                    # TODO it seems like val here is a string and not a list
+                    # TODO check case of 'host:' (empty set, all osds have it)
+                    if device in data['devices']:
+                        device_osds += osd
+            return (0, "", str(device_osds))
+
+        elif cmd['prefix'] == "smart warn":
+            # TODO check why it's not in 'ceph health' output
+            self.set_health_checks({
```
Comment (jcsp) [inline src/pybind/mgr/smart/module.py]: `set_health_checks` expects the complete list of health checks from this module.  So you need to accumulate all health checks that might be active, before calling this function.  You also need to call `set_health_checks({})` when there aren't any checks failing, to clear any that were raised before.

File: src/pybind/mgr/smart/module.py [hunk 2]
```diff
@@ -1,33 +1,561 @@
...
+    event = Event()
+    last_scrape_time = ""
+
+    def open_connection(self):
+        # TODO check that pool name is actually configurable
+        pool_name = self.get_config('pool_name') or 'smart_data'
+        # looking for the right pool
+        pools = self.rados.list_pools()
+        is_pool = False
+        for pool in pools:
```
Comment (tchaikov) [inline src/pybind/mgr/smart/module.py]: could just put
```py
if not pool_name in self.rados.list_pools():
  rados.create_pool(pool_name)
  self.log.debug('create %s pool' % pool_name)
# ...
```

File: src/pybind/mgr/smart/module.py [hunk 3]
```diff
@@ -1,33 +1,561 @@
...
+            self.log.debug('pool name is %s' % pool)
+            if pool == pool_name:
+                is_pool = True
+                break
+        if not is_pool:
+            self.rados.create_pool(pool_name)
+            self.log.debug('create %s pool' % pool_name)
+        ioctx = self.rados.open_ioctx(pool_name)
+        return (ioctx)
```
Comment (tchaikov) [inline src/pybind/mgr/smart/module.py]: no need to quote return value with parentheses.

File: src/pybind/mgr/smart/module.py [hunk 4]
```diff
@@ -1,33 +1,561 @@
...
+
+    def handle_osd_smart_scrape(self, osd_id):
+        ioctx = self.open_connection()
```
Comment (tchaikov) [inline src/pybind/mgr/smart/module.py]: as IoCtx also has the context manager support, we can simply put:
```py
with self.open_connection() as ioctx:
  self.do_osd_smart_scrape(osd_id, ioctx)
return (0, "", "")
```

File: src/pybind/mgr/smart/module.py [hunk 5]
```diff
@@ -1,33 +1,561 @@
...
+        self.do_osd_smart_scrape(osd_id, ioctx)
+        ioctx.close()
+        return (0, "", "")
+
+    # TODO maybe have one function that returns state tuple (in, up, ...)
+    def is_osd_in(self, osd_id):
+        osdmap = self.get("osd_map")
+        assert osdmap is not None
+        for osd in osdmap['osds']:
+            if str(osd_id) == str(osd['osd']) and osd['in']:
```
Comment (tchaikov) [inline src/pybind/mgr/smart/module.py]: should put
```py
if str(osd_id) == str(osd['osd']):
  return osd['in']
```

**Comment by tchaikov** [comment]
@yaarith @liewegas @jcsp is this feature still relevant or being actively worked on?

**Comment by liewegas** [comment]
Yep!  Let's leave this for now

**Comment by liewegas** [comment]
I've pulled parts of this into #22479 .. others will follow later.

**Pull request closed without merging**